corehq/apps/user_importer/importer.py

```diff
@@ -437,178 +437,182 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
+                else:
+                    web_user_importer = None
+                    user_change_logger = None
+                if remove_web_user:
+                    remove_web_user_from_domain(domain, web_user, username, upload_user,
+                                                user_change_logger)
+                else:
+                    check_user_role(username, role)
+                    if not web_user and is_account_confirmed:
+                        raise UserUploadError(_(
+                            "You can only set 'Is Account Confirmed' to 'True' on an existing Web User. "
+                            f"{web_user_username} is a new username."
+                        ).format(web_user_username=web_user_username))
+                    if web_user and not web_user.is_member_of(domain) and is_account_confirmed:
+                        # add confirmed account to domain
+                        # role_qualified_id would be be present here as confirmed in check_user_role
+                        web_user_importer.add_to_domain(role_qualified_id, user.location_id)
+                    elif not web_user or not web_user.is_member_of(domain):
+                        create_or_update_web_user_invite(web_user_username, domain, role_qualified_id,
+                                                        upload_user, user.location_id, user_change_logger,
+                                                        send_email=send_account_confirmation_email)
+                    elif web_user.is_member_of(domain):
+                        # edit existing user in the domain
+                        web_user_importer.update_role(role_qualified_id)
+                        if location_codes is not None:
+                            web_user_importer.update_primary_location(user.location_id)
+                        web_user.save()
+                if web_user_importer:
+                    web_user_importer.save_log()
+            if send_account_confirmation_email and not web_user_username:
+                send_account_confirmation_if_necessary(user)
+
+            if is_password(password):
+                # Without this line, digest auth doesn't work.
+                # With this line, digest auth works.
+                # Other than that, I'm not sure what's going on
+                # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
+                user.get_django_user(use_primary_db=True).check_password(password)
+
+            group_change_message = {}
+            if group_names:
+                group_change_message = commcare_user_importer.update_user_groups(domain_info, group_names)
+
+            try:
+                for group_id in domain_info.group_memoizer.updated_groups:
+                    group = domain_info.group_memoizer.groups_by_id[group_id]
+                    group.save()
+                domain_info.group_memoizer.updated_groups.clear()
```

Comment:
it would be great if this was encapsulated in the memoizer:
```python
def save_updated(self):
    updated = [self.groups_by_id[_id] for _id in self.updated_groups]
    Group.bulk_save(updated)
    self.updated_groups.clear()
```

Reply:
That makes sense. Just got around to finishing making this change. When I was working on this yesterday I realized that I needed to account for changes to groups alone. Previously group changes were dependent on a `log` being returned that could then be updated. That wasn't happening unless another change to the user was made first.